drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,831 @@
+	struct eeprom_emu_ctx ctx = {
+		.data = data,
+		.len = len,
+		.address = address,
+		.rlen = len,
+	};
+	int rc = 0;
+
+	/* Nothing to do */
+	if (!len) {
+		return 0;
+	}
+
+	/* Error checking */
+	if ((!data) || (!eeprom_emu_range_is_valid(dev, address, len))) {
+		return -EINVAL;
+	}
+
+	if (dev_config->readonly) {
+		LOG_ERR("attempt to write to read-only device");
+		return -EACCES;
+	}
+
+	if (!device_is_ready(dev_config->flash_dev)) {
+		LOG_ERR("flash device is not ready");
+		return -EIO;
+	}
+
+	/* Handle normal case */
+	LOG_DBG("EEPROM write at [0x%tx] length[%d]", (ptrdiff_t)address, len);
+
+	k_mutex_lock(&dev_data->lock, K_FOREVER);
+
+	/* first update the rambuffer */
+	if (dev_config->rambuf) {
+		memcpy(dev_config->rambuf + address, data, len);
+	}
+
+	/* second update the flash */
+	while (ctx.rlen) {
+		rc = eeprom_emu_flash_set(dev, &ctx);
+		if (rc) {
+			break;
+		}
+
+	}
+
+	k_mutex_unlock(&dev_data->lock);
+
+	return rc;
+}
+
+static size_t eeprom_emu_size(const struct device *dev)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	return dev_config->size;
+}
+
+static int eeprom_emu_init(const struct device *dev)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	off_t offset;
+	uint8_t buf[dev_config->flash_cbs];
+	int rc;
+
+	k_mutex_init(&dev_data->lock);
+	if (!device_is_ready(dev_config->flash_dev)) {
+		__ASSERT(0, "Could not get flash device binding");
+		return -ENODEV;
+	}
+
+	/* Find the page offset */
+	dev_data->page_offset = 0U;
+	dev_data->write_offset = dev_config->page_size - sizeof(buf);
+	while (dev_data->page_offset < dev_config->flash_size) {
+		offset = dev_data->page_offset + dev_data->write_offset;
+		rc = eeprom_emu_flash_read(dev, offset, buf, sizeof(buf));
+		if (rc) {
+			return rc;
+		}
+
+		if (!eeprom_emu_is_word_used(dev, buf)) {
+			break;
+		}
+
+		dev_data->page_offset += dev_config->page_size;
+	}
+
+	if (dev_data->page_offset == dev_config->flash_size) {
+		__ASSERT(0, "All pages are invalid, is this a EEPROM area?");
+		return -EINVAL;
+	}
+
+	dev_data->write_offset = dev_config->size;
+
+	/* Update the write offset */
+	while ((dev_data->write_offset + sizeof(buf)) < dev_config->page_size) {
+		offset = dev_data->page_offset + dev_data->write_offset;
+		rc = eeprom_emu_flash_read(dev, offset, buf, sizeof(buf));
+		if (rc) {
+			return rc;
+		}
+
+		if (!eeprom_emu_is_word_used(dev, buf)) {
+			break;
+		}
+
+		dev_data->write_offset += sizeof(buf);
+	}
+
+	/* dev_data->write_offset reaches last possible location, compaction
+	 * might have been interrupted: call eeprom_emu_compactor again, but
+	 * only in case we are using a write-enabled eeprom
+	 */
+	if ((!dev_config->readonly) &&
+	    ((dev_data->write_offset + sizeof(buf)) >= dev_config->page_size)) {
+		/* Need to write in next step, disable flash write protection */
+		if (flash_write_protection_set(dev_config->flash_dev, false)) {
```

Comment:
flash_write_protection_set(..., flase) in eeprom_emu_flash_write/erase are sufficient. No need to do it here.

Reply:
Oops, forgot to remove in the init routine...